theano/gpuarray/ctc.py

```diff
@@ -0,0 +1,174 @@
+
+        self.compute_grad = compute_grad
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
+
+        gof.COp.__init__(self, self.func_file, self.func_name)
+
+    def c_lib_dirs(self):
+        lib_dirs = []
+        if ctc_available.path is not None:
+            lib_dirs += [ctc_available.path]
+        return lib_dirs
+
+    def c_libraries(self):
+        return ["warpctc", "gpuarray"]
+
+    def c_header_dirs(self):
+        dirs = [os.path.dirname(__file__), pygpu.get_include()]
+        if config.ctc.root != '':
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        return dirs
+
+    def c_headers(self):
+        return ['ctc.h', 'numpy_compat.h', 'gpuarray/ext_cuda.h',
+                'gpuarray_helper.h', 'gpuarray/types.h', 'gpuarray_api.h',
+                'gpuarray/array.h', 'gpuarray/util.h', 'gpuarray/extension.h']
+
+    def get_params(self, node):
+        return node.inputs[0].type.context
+
+    def make_node(self, activations, labels, input_lengths):
+        context_name = infer_context_name(activations)
+        t_activations = as_gpuarray_variable(activations,
+                                             context_name=context_name)
+        # Ensure activations array is C-contiguous
+        t_activations = gpu_contiguous(t_activations)
+
+        # Labels and input lengths are always on the CPU
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.as_tensor_variable(input_lengths)
+
+        if t_activations.type.dtype != 'float32':
+            raise TypeError('activations must use the float32 type.')
+
+        if t_activations.ndim != 3:
+            raise ValueError('activations must have 3 dimensions.')
+
+        if t_labels.type.dtype != 'int32':
+            raise TypeError('labels must use the int32 type.')
+
+        if t_labels.ndim != 2:
+            raise ValueError('labels must have 2 dimensions.')
+
+        if t_input_lengths.type.dtype != 'int32':
+            raise TypeError('input_lengths must use the int32 type.')
+
+        if t_input_lengths.ndim != 1:
+            raise ValueError('input_lengths must have 1 dimension.')
+
+        costs = GpuArrayType(dtype='float32',
+                             broadcastable=(False,),
+                             context_name=context_name)()
+        outputs = [costs]
+
+        if self.compute_grad:
+            gradients = GpuArrayType(dtype='float32',
+                                     broadcastable=(False, False, False,),
+                                     context_name=context_name)()
+            outputs += [gradients]
+
+        return theano.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                            outputs=outputs)
+
+    def L_op(self, inputs, outputs, output_grads):
+        # Gradients computed by Op
+        gradients = outputs[1]
+        assert gradients is not None
```

Comment:
I've added the checks on both the length and the compute_grad flag on both implementations.